dbms/src/TiDB/Schema/SchemaBuilder.cpp

```diff
@@ -219,6 +219,24 @@ void SchemaBuilder<Getter, NameMapper>::applyDiff(const SchemaDiff & diff)
         applyPartitionDiff(diff.schema_id, diff.table_id);
         break;
     }
+    case SchemaActionType::ActionAlterTablePartitioning:
+    {
+        if (diff.table_id == diff.old_table_id)
```

Comment:
It is depending on the stage change, only the last (StateDeleteReorganization -> StatePublic) is the table ID changed to the new one.

Example (added custom logs):
tidb.log
```
[2023/08/01 23:11:59.568 +02:00] [WARN] [ddl_worker.go:1431] [updateSchemaVersion] [category=ddl] [TableId=96] [SchemaState="delete only"]
[2023/08/01 23:11:59.637 +02:00] [WARN] [ddl_worker.go:1431] [updateSchemaVersion] [category=ddl] [TableId=96] [SchemaState="delete only"]
[2023/08/01 23:11:59.702 +02:00] [WARN] [ddl_worker.go:1431] [updateSchemaVersion] [category=ddl] [TableId=96] [SchemaState="write only"]
[2023/08/01 23:11:59.798 +02:00] [WARN] [ddl_worker.go:1431] [updateSchemaVersion] [category=ddl] [TableId=96] [SchemaState="write reorganization"]
[2023/08/01 23:11:59.865 +02:00] [WARN] [ddl_worker.go:1431] [updateSchemaVersion] [category=ddl] [TableId=96] [SchemaState="delete reorganization"]
[2023/08/01 23:11:59.865 +02:00] [WARN] [ddl_worker.go:1441] [StateDeleteReorganization] [category=ddl] [TableId=105] [OldTableID=96]
```
tiflash.log
```
[2023/08/01 23:11:59.818 +02:00] [WARN] [SchemaBuilder.cpp:238] ["ActionRemovePartitioning table_id 96 old 96"] [source="keyspace=4294967295"] [thread_id=60]
[2023/08/01 23:11:59.819 +02:00] [WARN] [SchemaBuilder.cpp:238] ["ActionRemovePartitioning table_id 96 old 96"] [source="keyspace=4294967295"] [thread_id=60]
[2023/08/01 23:11:59.820 +02:00] [WARN] [SchemaBuilder.cpp:238] ["ActionRemovePartitioning table_id 96 old 96"] [source="keyspace=4294967295"] [thread_id=60]
[2023/08/01 23:11:59.821 +02:00] [WARN] [SchemaBuilder.cpp:238] ["ActionRemovePartitioning table_id 96 old 96"] [source="keyspace=4294967295"] [thread_id=60]
[2023/08/01 23:12:55.086 +02:00] [WARN] [SchemaBuilder.cpp:238] ["ActionRemovePartitioning table_id 105 old 96"] [source="keyspace=4294967295"] [thread_id=61]
```

Reply:
So in the first three stages, does the partition.definitions changes?

Reply:
In the first (StateNone->StateDeleteOnly), it adds the new partitions to TableInfo.Partition.AddingDefinitions and the removed ones to TableInfo.Partition.DroppingDefinitions.

In the second (StateDeleteOnly->StateWriteOnly), it only waits for the partitions to be available in TiFlash (if the table should have any TiFlash replicas).

In the third (StateWriteOnly->StateWriteReorganization), no changes or checks.

In the forth (StateWriteReorganization->StateDeleteReorganization), no changes (but does all the data copying and recreating the indexes).

In the fifth (StateDeleteReorganization->StatePublic), it depends on the actual command (reorganize, remove partitioning or alter table partition by):
- Reorganize, it will modify the TableInfo.Partition.Definitions by removing the DroppingDefinitions and add the AddingDefinitions and then reset AddingDefinitions and DroppingDefinitions.
- Remove partitioning, it will change the tableID to the one used for AddingDefinitions. Really drop the old partitioned table metadata, change the TableInfo by setting the table id to the single AddingDefinitions table id, and remove TableInfo.Partition, and then create the table (metadata) with the updated TableInfo.
- Alter Table Partition By, it will change the TableID and replace the TableInfo.Partition with the new partitioning type, expression and definitions. Also by first drop the table (metadata), change the TableInfo.Partition and table id, and the (re)create the table with the updated TableInfo.

Notice that the data cleanup is done later in an async way, and that the old table data also needs to be accessible during the schemaversion the DDL got during transitioning to StatePublic, since there may be clients in the schemaversion of StateDeleteReorganization that still need to read and (double) write the data.

Reply:
After some additional testing I figured out that I missed to drop the table (metadata) in TiFlash as well, not just create a new table :)